File: src/main/java/victor/training/cleancode/refactoring/ExtractMethodObject.java
package victor.training.cleancode.refactoring;

import jakarta.persistence.EntityManager;
import jakarta.persistence.TypedQuery;
import lombok.RequiredArgsConstructor;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RequiredArgsConstructor
public class ExtractMethodObject {
  private final EntityManager em;

  public List<Long> search(CustomerSearchCriteria criteria) {
    String jpql = "SELECT c.id FROM Customer c WHERE 1=1 ";
    Map<String, Object> params = new HashMap<>();

    if (criteria.name != null) {
      jpql += " AND UPPER(c.name) LIKE '%' || UPPER(:name) || '%' ";
      params.put("name", criteria.name);
    }

    if (criteria.countryId != null) {
      jpql += " AND (c.residenceCountry.id = :countryId OR ..<5 lines of JPQL>..)";
      params.put("countryId", criteria.countryId);
    }

    TypedQuery<Long> query = em.createQuery(jpql + " ", Long.class);
    for (String param : params.keySet()) {
      query.setParameter(param, params.get(param));
    }
    return query.getResultList();
  }
}

class CustomerSearchCriteria {

  public String name;
  public Long countryId;
}

File: src/main/java/victor/training/cleancode/ExtractMethodObject.kt
package victor.training.cleancode

import lombok.RequiredArgsConstructor
import org.apache.commons.lang.StringUtils
import javax.persistence.EntityManager

@RequiredArgsConstructor
class ExtractMethodObject {
    private val em: EntityManager? = null
    fun search(criteria: CustomerSearchCriteria): List<Long> {
        return SearchMethodObject(criteria,em!!).execute()
    }
}

class CustomerSearchCriteria {
    var name: String? = null
    var countryId: Long? = null
}